packages/ampd-sdk/src/event/event_handler.rs

```diff
@@ -0,0 +1,563 @@
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct Config {
+    #[serde(with = "humantime_serde")]
+    pub stream_timeout: Duration,
+}
+
+impl Default for Config {
+    fn default() -> Self {
+        Self {
+            stream_timeout: Duration::from_secs(30),
+        }
+    }
+}
+
+#[derive(Error, Debug)]
+pub enum Error {
+    #[error("failed to retreive events stream from the client")]
+    ClientEventStream,
+
+    #[error("timeout while waiting for event stream")]
+    StreamTimeout,
+
+    #[error("unable to parse event of type")]
+    EventConversion,
+
+    #[error("error when handling event messages")]
+    HandlerFailed,
+
+    #[error("error when broadcasting message")]
+    BroadcastFailed,
+
+    #[error("{0} task aborted")]
+    TaskAborted(String),
+}
+
+pub enum HandlerTaskAction {
+    Abort,
+    Continue,
+    ContinueAndSleep(Duration),
+}
+
+#[allow(dead_code)]
+#[derive(TypedBuilder)]
+pub struct HandlerTask<H>
+where
+    H: EventHandler,
+    <H::Event as TryFrom<Event>>::Error: Context,
+    H::Event: Display,
+{
+    handler: H,
+    config: Config,
+    #[builder(default = RetryPolicy::NoRetry)]
+    handler_retry_policy: RetryPolicy,
+    #[builder(default = RetryPolicy::NoRetry)]
+    broadcast_retry_policy: RetryPolicy,
+    #[builder(default = HandlerTask::<H>::default_event_subscription_error_cb)]
+    event_subscription_error_cb: fn(Report<Error>) -> HandlerTaskAction,
+    #[builder(default = HandlerTask::<H>::default_handler_error_cb)]
+    handler_error_cb: fn(&H::Event, H::Err) -> HandlerTaskAction,
+    #[builder(default = HandlerTask::<H>::default_broadcast_error_cb)]
+    broadcast_error_cb: fn(&[Any], H::Err) -> HandlerTaskAction,
+}
+
+impl<H> HandlerTask<H>
+where
+    H: EventHandler,
+    <H::Event as TryFrom<Event>>::Error: Context,
+    H::Event: Display,
+{
+    fn default_event_subscription_error_cb(err: Report<Error>) -> HandlerTaskAction {
+        error!(
+            err = LoggableError::from(&err).as_value(),
+            "failed to subscribe to events"
+        );
+
+        HandlerTaskAction::Continue
+    }
+
+    fn default_handler_error_cb(event: &H::Event, err: H::Err) -> HandlerTaskAction
+    where
+        <H::Event as TryFrom<Event>>::Error: Context,
+        H::Event: Display,
+    {
+        error!(
+            err = LoggableError::from(&report!(err)).as_value(),
+            event = format!("{event}").as_value(),
+            "failed to process events"
+        );
+
+        HandlerTaskAction::Continue
+    }
+
+    fn default_broadcast_error_cb(msgs: &[Any], err: H::Err) -> HandlerTaskAction {
+        let msgs = msgs
+            .iter()
+            .map(serde_json::to_string)
+            .try_collect()
+            .unwrap_or(vec![]);
+
+        error!(
+            err = LoggableError::from(&report!(err)).as_value(),
+            msgs = msgs.as_value(),
+            "failed to broadcast messages"
+        );
+
+        HandlerTaskAction::Continue
+    }
```

Comment:
Don't think these need to be methods in the handler task itself but individual functions. Also, I'd move them to some other file.

Reply:
oh yeah, didn't realize I could make the associated types generics on the functions to move them out

Reply:
They shouldn't need to. They should only know what these callbacks are for and what their signatures are. Good interface design should eliminate the need for the user to understand what's going on behind the scenes.